### PR TITLE
Data on slice for DataBox

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

1. Adds a `data_on_slice` function that is similar to the current `Variables` version but retrieves `Tensor`s from the DataBox rather than slicing a `Variables`. This will be useful for slicing primitives and auxiliary variables such as the lapse, shift, etc. for GH.
2. Move `DataBox.hpp` and friends into `src/DataStructures/DataBox` as was planned a while back.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: src/DataStructures/DataBox/DataOnSlice.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "DataStructures/DataBox.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/SliceIterator.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+namespace db {
+template <size_t VolumeDim, typename TagsList, typename... TagsToSlice>
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox/DataOnSlice.hpp]: add documentation

**Comment by nilsdeppe** [comment]
@kidder could you look at this?

@wthrowe I added the doxygen documentation we had discussed in our last review. Let me know if should be improved :)

Rendered documentation is available [here](http://nilsdeppe.com/spectre/pr_474/)

**Comment by nilsdeppe** [comment]
rebased on develop, we'll need to make sure that this is fully rebased at time of merge since it moves files around and that doesn't necessarily show up as a git conflict.

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
because of huge queue for mac builds on Travis, I tested this on my laptop

**Pull request merged**